File: solutions/python/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if not obstacleGrid or not obstacleGrid[0] or obstacleGrid[0][0] == 1:
            return 0 
        
        rows, cols = len(obstacleGrid), len(obstacleGrid[0])
        
        previous = [0] * cols
        current = [0] * cols
        previous[0] = 1 
        
        for r in range(rows):
            current[0] = 0 if obstacleGrid[r][0] == 1 else previous[0]            
            for c in range(1, cols):
                current[c] = 0 if obstacleGrid[r][c] == 1 else current[c-1] + previous[c]
            
            previous[:] = current
        
        return previous[cols-1]

File: solutions/go/0063-unique-paths-ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    if len(obstacleGrid) == 0 || len(obstacleGrid[0]) == 0 || obstacleGrid[0][0] == 1 {
        return 0
    }

    m := len(obstacleGrid)
    n := len(obstacleGrid[0])

    previous := make([]int, n)
    current := make([]int, n)
    previous[0] = 1

    for i := 0; i < m; i++ {
        if obstacleGrid[i][0] == 1 {
            current[0] = 0
        } else {
            current[0] = previous[0]
        }
        
        for j := 1; j < n; j++ {
            if obstacleGrid[i][j] == 1 {
                current[j] = 0
            } else {
                current[j] = current[j-1] + previous[j]
            }
        }
        previous, current = current, previous
    }
    return previous[n-1]
}
